core/src/main/java/com/linecorp/armeria/common/util/CachingSupplier.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.util;
+
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.client.limit.ConcurrencyLimit;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * Supplies the cached value that is useful below cases.
+ * <p>
+ * - Supplies dynamically a maximum number of concurrent active requests for {@link ConcurrencyLimit}.
+ */
+public class CachingSupplier<T> implements Supplier<T> {
+    @Nullable
+    private volatile T value;
+
+    /**
+     * Creates a new instance with the specified {@code initialValue}.
+     */
+    public static <T> CachingSupplier<T> of(@Nullable T initialValue) {
+        return new CachingSupplier<>(initialValue);
+    }
+
+    protected CachingSupplier(@Nullable T initialValue) {
```

Comment:
Question:
I'm not sure if it's a good idea but have you thought about making this class final which is not inheritable?

Reply:
I thought it would be useful to inherit as below codes.
```java
class DynamicSupplier extends SettableSupplier<Integer> {
    public DynamicSupplier() {
        super(10); //set initial value
        AnyListener<Integer> listener = ...
        listener.addListener(this::set);
    }
}
```

We had explained the example in `ConcurrencyLimit` class as below.
I think it is not good way.
This example does not inherit `SettableSupplier`, but use `SettableSupplier` as member variable. It looks `DynamicLimit` supplier supply a value using other supplier(`SettableSupplier`).
But we can't explain everything in `ConcurrencyLimit` class until `SettableSupplier`.
So the example is not good.
```java
class DynamicLimit implements Supplier<Integer> {
    private final SettableSupplier<Integer> settableSupplier = SettableSupplier.of(16);
    public DynamicLimit() {
        LimitChangeListener<Integer> listener = ...
        listener.addListener(updatedValue -> settableSupplier.set(updatedValue));
    }
    @Override
    public final Integer get() {
        return settableSupplier.get();
    }
}
```

What do you think about adding an example of inheritance in `SettableSupplier`?

Reply:
Thanks for your opinion. 😄 
Yeah, we don't have to stick to the composition in this case because `DynamicLimit` is just created and passed to the `ConcurrentLimit` so `set(...)` method is not exposed anyway.
So either example is fine to me. 😄